File: C/mwt.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

typedef struct Node Node;

struct Node {
    int *data;
    Node **child;
    Node *parent;
    int numOfValues;
};

//create node
Node * createTree(char *inputStream, int fileLength, int **alphabet, int arity, int numOfDigits){

    Node* root = (Node *) malloc(sizeof(Node));
    (*root).data = (int *)malloc(fileLength*sizeof(int));
    (*root).data[0] = -1;
    (*root).child = (Node **)malloc(arity*sizeof(Node*));
    (*root).parent = NULL;
    (*root).numOfValues = 0;
    int i;
    for(i = 0; i < arity; i++){
        Node* child = (Node *)malloc(sizeof(Node));
        (*child).data = (int *)malloc(sizeof(int));
        (*child).data[0] = -1;
        (*child).parent = root;
        (*root).child[i] = child;
    }

    //for each character in input stream
    for (i = 0; i < fileLength; i++){
        int c = inputStream[i];
        int *codeValue = alphabet[c];

        //save each digit to appropriate node
        int j;
        Node* temp = root;
        for(j = 0; j < numOfDigits; j++){
            int digit = codeValue[j];
            //put first digit in root
            if(j==0){
                (*root).data[i] = digit;
                (*root).numOfValues = (*root).numOfValues + 1;
            }
            else {
                  Node* node = (*temp).child[codeValue[j-1]];
                    //create child if doesn't exist
                if((*node).data[0] == -1){
                    (*node).data[0] = digit;
                    (*node).numOfValues = 1;
                    (*node).child = (Node **)malloc(arity*sizeof(Node*));
                    (*node).parent = (Node *)malloc(sizeof(Node));
                    int k;
                    for(k = 0; k < arity; k++){
                        Node* children = (Node *)malloc(sizeof(Node));
                        (*children).data = (int *)malloc(sizeof(int));
                        (*children).data[0] = -1;
                        (*children).parent = node;
                        (*node).child[k] = children;
                    }
                    temp = node;
                }
                //child already exists
                else {
                    int size = (*node).numOfValues + 1;
                    (*node).data = (int *) realloc((*temp).child[codeValue[j-1]]->data, size*sizeof(int));
                    (*node).data[size-1] = digit;
                    (*node).numOfValues = (*node).numOfValues + 1;
                    temp = node;
                }
            }
            }
    }
    return root;
}

//convert number to different base; result is code value of alphabet character
int * calculateCodeValue (int value, int base, int numOfDigits){
    int *codeValue = (int *) malloc(numOfDigits*sizeof(int));

    if(base==10 || value==0){
        codeValue[numOfDigits - 1] = value;
    }
    else {
        int result = value;
        int modResult = 0;
        int i = 1;
        while(result != 0){
            modResult = result % base;
            codeValue[numOfDigits - i] = modResult;
            i++;
            result = result/base;
        }
        while(i < (numOfDigits+1)){
            codeValue[numOfDigits - i] = 0;
            i++;
        }
    }

    return codeValue;
}

int main (int argc, char *argv[]){
    if(argc != 3){
        printf("Number of parameters should be 2 (filename, arity).");
        exit(1);
    }

    int *alphabet[128] = {NULL};
    int arity = atoi(argv[2]);
    int treeLayers = ceil(7/log2(arity));
    //printf("Number of layers: %d \n", treeLayers);

    //open file and get file length
    FILE *file;
    file = fopen(argv[1],"r");
    if(file == NULL){
        printf("File couldn't be opened.");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    int fileLength = ftell(file);
    fseek(file, 0, SEEK_SET);

    //save data from file to array
    char inputStream[fileLength];
    printf("length: %d \n", fileLength);
    int i = 0;
    int character;
    while((character = fgetc(file)) != EOF){
            inputStream[i] = character;
            if(alphabet[character]==NULL){
                alphabet[character] = calculateCodeValue(character, arity, treeLayers);
            }
            i++;
            }
    fclose(file);

    Node * root = createTree(&inputStream[0], fileLength, &alphabet[0], arity, treeLayers);

    printf("Creating tree finished.\n");
    int a = 1;
    while (a) {
        char command;
        int position;
        char charOfAlphabet;
        int rank, select;
        printf("Continue (c) or exit (e)? \n");
        scanf(" %c", &command);
        if (command=='e'){
            a = 0;
            break;
        }
        else if(command=='c'){
            printf("\nType your command.\n\n");
            printf("Characters in the tree:\n");
            for(j = 0; j < 128; j++){
                if(alphabet[j] != NULL){
                    printf("%c ", j);
                }
            }
            printf("\n\n");
            printf("Possible commands: \n \n");
            printf("1) RANK - r 'position' 'character' \t \t MAX_POSITION = %d", fileLength);
            printf("\n \n");
            printf("2) SELECT - s 'num_of_character' 'character' \n \n");
            scanf(" %c %d %c", &command, &position, &charOfAlphabet);
            if(command=='r'){
                if(position > fileLength || position < 0){
                    printf("Wrong value of rank.\n\n");
                }
                else if(alphabet[charOfAlphabet] == NULL){
                    printf("There is no such character in alphabet.\n\n");
                }
                else {
                    int *codeValue = alphabet[charOfAlphabet];
                    //rank
					//printf("Calculated rank: %d", rank);
                }
            }
            else if(command=='s'){
                if(alphabet[charOfAlphabet] == NULL){
                    printf("There is no such character in alphabet.\n\n");
                }
                else {
                    //select
                    //printf("%d. character %c is on position: %d", position, charOfAlphabet, select);
                }
            }
            else {
                printf("Wrong command.\n");
            }
        }
        else {
            printf("Wrong command. Type 'c' for continue or 'e' for exit.\n\n");
        }
    }

return 0;
}

File: java/mwt.java
import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Scanner;


public class mwt {
	
	public static HashMap<Character,ArrayList<Integer>> alphabet = new HashMap<Character,ArrayList<Integer>>();
	
	public static ArrayList<Integer> CalculateIntoBase(int ascii_character, int cardinality, int number_layer) {
		ArrayList<Integer> number_in_list = new ArrayList<Integer>();
		int div, mod;
		
		div = ascii_character/cardinality;
		mod = ascii_character%cardinality;
		number_in_list.add(mod);
		
			while(div > 0) {
				mod = div%cardinality;
				div = div/cardinality;
				number_in_list.add(mod);
				
			}
			//reverse array
			Collections.reverse(number_in_list);
			
			//fill lists with 0s
			while(number_in_list.size() != number_layer) {
				number_in_list.add(0, 0);
			}
		
		return number_in_list;
	}
	
	//rank(position, symbol)
	public static int Rank(Root root, int position, char character) {
		ArrayList<Integer> char_into_base = new ArrayList<Integer>();
		
		int digit, rank = 0;
		Root rootNode = root;
		
		char_into_base.addAll(alphabet.get(character));
		
		digit = char_into_base.remove(0);
		rank = Rank_layer(rootNode, position, digit);
		Node node = rootNode.children.get(digit);
		position = rank;
		
		while(char_into_base.size() > 0) {
			digit = char_into_base.remove(0);
			rank = Rank_layer(node, position, digit);
			if(char_into_base.size() == 0) {
				break;
			}
			
			node = node.children.get(digit);
			position = rank;
		}
		char_into_base.clear();
		return rank;
	}
	
	//for each layer returns the number of symbol in data
	public static int Rank_layer(Root node, int position, int digit) {
		int rank = 0, i;
		ArrayList<Integer> data = node.getData();
		
		for(i = 0; i < position; i++) {
			if(data.get(i) == digit) {
				rank++;
			}
		}
		
		return rank;
	}
	
	//select(rank, symbol)
	public static int Select (Root root, int num_repeat, char character) {
		ArrayList<Integer> char_into_base = new ArrayList<Integer>();
		int i, digit, select = 0;
		
		Root rootNode = root;
		Node node;
		boolean temp = false;
		
		char_into_base.addAll(alphabet.get(character));
		
		digit = char_into_base.get(0);
		node = rootNode.children.get(digit);
		
		for(i = 0; i < char_into_base.size()-2; i++) {
			digit = char_into_base.get(i+1);
			node = node.children.get(digit);
		}
		
		Collections.reverse(char_into_base);
		
		while(char_into_base.size() > 0) {
			digit = char_into_base.remove(0);
			//nodes
			if(!temp) {
				select = Select_layer(node, num_repeat, digit);
			}
			//root nodes
			else{
				select = Select_root(root, num_repeat, digit);
			}
			
			if(select == -1) {
				System.out.println("Error!");
				break;
			}
			
			if(char_into_base.size() == 0) {
				break;
			}
			
			if(node.parent == null) {
				root = node.parent_root;
				temp = true;	
			}
			else {
				node = node.parent;
			}
			
			num_repeat = select + 1;
			
		}
		char_into_base.clear();
		return select;
	}
	
	public static int Select_layer(Node node, int num_repeat, int digit) {
		int select = 0, i;
		ArrayList<Integer> data = node.getData();
		
		if(num_repeat > data.size()) {
			return -1;
		}
		
		for(i = 0; i < data.size(); i++) {
			if(data.get(i) == digit) {
				select++;
			
			if (select == num_repeat) {
				return i;
			}
		}
	}
		return -1;
}
	//when node is root
	public static int Select_root(Root root, int num_repeat, int digit) {
		int select = 0, i;
		ArrayList<Integer> data = root.getData();
		
		for(i = 0; i < data.size(); i++) {
			if(data.get(i) == digit) {
				select++;
			
			if (select == num_repeat) {
				return i;
			}
		}
	}
		return -1;
}
	
	public static void main(String[] args) throws IOException {
		char i;
		int j;
		
		if(args.length != 2) {
			System.err.println("Invalid command line, exactly two argument required!");
			System.exit(1);
		}
		
		int cardinality = Integer.parseInt(args[1]);
		String filename = args[0];
		File file = new File(filename);
		
		try {
			
			//time to executed method
			long startTime = System.currentTimeMillis();
			
			//open file
			Scanner fileScanner = new Scanner(file);
			String queue = fileScanner.nextLine();
			
			ArrayList<Integer> number_base = new ArrayList<Integer>();
			
			//number of layers
			int number_layer = (int)Math.ceil((7/(Math.log(cardinality)/Math.log(2))));
			Root root = new Root();
			
			queue.trim();
			
			//read character and if doesn't exist save in alphabet
			for(j = 0; j < queue.length(); j++) {
				i = queue.charAt(j);
				if(!alphabet.containsKey(i)) {
					//calculate base of a character
					number_base = CalculateIntoBase((int)i, cardinality, number_layer);
					alphabet.put(i, number_base);
				}
				
				root.Add_signs(alphabet.get(i));
			}
			System.out.println("Creating tree finished!");
			
			
			long endTime = System.currentTimeMillis();
			System.out.println("Time: " + (endTime - startTime) + " milliseconds");
			
			Runtime runtime_max = Runtime.getRuntime();
		 	Runtime runtime_free = Runtime.getRuntime();
			
		 	double memoryUsage = (double)runtime_max.totalMemory()/1024 - (double)runtime_free.freeMemory()/1024;
			System.out.println("Memory usage: " + memoryUsage + " KB");
			
			int position, _rank;
			char symbol;
			boolean temp = true;
			BufferedReader input_position, input_symbol;
			long startTimeCommand, endTimeCommand;
			Runtime runtime_max_command, runtime_free_command;
			double memoryUsage_command;
			
			while(temp) {
				BufferedReader functionInput = new BufferedReader(new InputStreamReader(System.in));
				System.out.print("Enter command <rank/select/exit>: ");
				String function = functionInput.readLine();
				if(function.equals("rank")) {
					
					 input_position = new BufferedReader(new InputStreamReader(System.in));
					 System.out.print("Enter position: ");
					 position = Integer.parseInt(input_position.readLine());
					 
					 input_symbol = new BufferedReader(new InputStreamReader(System.in));
					 System.out.print("Enter symbol: ");
					 symbol = input_symbol.readLine().charAt(0);
					
					 startTimeCommand = System.currentTimeMillis();
					 
					 int rank = Rank(root, position, symbol);
					 System.out.println("Rank = " + rank);
					 
					 endTimeCommand = System.currentTimeMillis();
					 System.out.println("Time: " + (endTimeCommand - startTimeCommand) + " milliseconds");
						
					 runtime_max_command = Runtime.getRuntime();
					 runtime_free_command = Runtime.getRuntime();
						
					 memoryUsage_command = (double)runtime_max_command.totalMemory()/1024 - (double)runtime_free_command.freeMemory()/1024;
					 System.out.println("Memory usage: " + memoryUsage_command + " KB");		
			}
				if(function.equals("select")) {
					
					 input_position = new BufferedReader(new InputStreamReader(System.in));
					 System.out.print("Enter position: ");
					 _rank = Integer.parseInt(input_position.readLine());
					 
					 input_symbol = new BufferedReader(new InputStreamReader(System.in));
					 System.out.print("Enter symbol: ");
					 symbol = input_symbol.readLine().charAt(0);
					
					 startTimeCommand = System.currentTimeMillis();
					 
					 int select = Select(root, _rank, symbol);
					 if(select == -1) {
						 System.out.println("There aren't that many symbols in input file");
					 }
					 else {
						 System.out.println(_rank + ". " + "character " + symbol + " is on position: " + select);
						 
						 endTimeCommand = System.currentTimeMillis();
						 System.out.println("Time: " + (endTimeCommand - startTimeCommand) + " milliseconds");
							
						 runtime_max_command = Runtime.getRuntime();
						 runtime_free_command = Runtime.getRuntime();
							
						 memoryUsage_command = (double)runtime_max_command.totalMemory()/1024 - (double)runtime_free_command.freeMemory()/1024;
						 System.out.println("Memory usage: " + memoryUsage_command  + " KB");
					 }
					 
			}
				else if(function.equals("exit")) {
					temp = false;
					break;
				}
		}
			fileScanner.close();
		}
		
		finally {
			
		}
	}
}